### PR TITLE
Helpful error when constant not found

### LINKED ISSUE #6122: Beginner friendly 'require' message when a constant is not found, but it is in the codebase
> Relates to https://github.com/crystal-lang/crystal/issues/5291 I started this new issue because the other one was already quite long. 

```crystal
# src/send_email/invitation_email.cr
class InvitationEmail
  include Mailer
  # ...code
end

# src/send_email/mailer.cr
module Mailer
  # ...code
end

# src/send_email.cr
require "./emails/*
InvitationEmail.new.send
```

This will fail to find the constant `Mailer` when you run `crystal src/send_email.cr` because the files are loaded alphabetically. Since `InvitationEmail` is loaded *before* the required `Mailer` dependency it fails. This may be unexpected for newcomers if they are using `require "*"` and it works fine until they happen to have a dependency that due naming is not loaded yet.

I think it would be helpful for the compiler to check the `src` folder for unrequired files that have a class/module that matches and tell the user what to require to get it working. In this case:

```crystal
Could not find module 'Mailer' in 'src/send_email/invitiation_email.cr'

Maybe you need to require it first? 

   # in src/send_email/invitation_email.cr
   require "./mailer"
```

It wouldn't find everything (for example if you forgot to load "colorize" and tried to use the `colorize` method), but I think this would help a lot for some of the common cases.

Note, I am not saying the compiler actually compiles the unrequired file, it just does a simple text search looking for the constant in `.cr` files so it can see if you forgot to require one.

### Challenges

Constants nested in modules like this would be easy to find:

```crystal
class MyNamespace::MyClass
end
```

But I imagine this would be harder

```crystal
module MyNameSpace
  class MyClass
  end
end
```

### Different solutions?

I'm not sure this is the *best* solution, but this is definitely a problem I've seen a number of times that trips people up. I'd love to hear alternatives if this won't work or would be too hard

### PR DESCRIPTION
**Pull request by paulcsmith**
Closes #6122 

screen shot 2018-05-30 at 9 51 40 pm

### CODE REVIEW

**Comment by paulcsmith** [comment]
A few failing specs where I need to use the `contains` matcher to handle the new error messages. Will tackle that sometime this week

File: src/compiler/crystal/semantic/call_error.cr [hunk 1]
```diff
@@ -28,10 +28,20 @@ class Crystal::Path
...
     else
-      self.raise("undefined constant #{self}")
+      example_require = %(require "./#{self.to_s.underscore}")
+      colorized_require_example = type.program.colorize(example_require).yellow.bold
+      self.raise <<-ERROR
+      undefined constant #{type.program.colorize(self).bold.yellow}
+
+      Try this...
+
+        ▸ Require the file with #{type.program.colorize(self).bold}. For example: #{colorized_require_example}
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/call_error.cr]: *the file with Foo* seems a bit imprecise. What about *the file that defines Foo*?
Or, more similar to the second suggestion:
> If Foo is declared in a file, add a relative require `require "./foo"`

Comment (paulcsmith) [reply src/compiler/crystal/semantic/call_error.cr]: Great suggestion 👍

**Comment by paulcsmith** [comment]
I have been working on some other stuff with Lucky, but I'll come back to this PR this week (hopefully!).

**Comment by j8r** [comment]
@paulcsmith what a long week :smile: Have  you the time do continue on this?

**Comment by straight-shoota** [comment]
Closing stale PR. There has been no activity for a long time and there's no point in keeping this open.
The feature enhancement is still tracked in #6122.

**Pull request closed without merging**